chainer/computational_graph.py

```diff
@@ -0,0 +1,106 @@
+import heapq
+
+from chainer import function
+from chainer import variable
+
+
+class DotNode(object):
+    def _shape(self):
+        if isinstance(self.node, variable.Variable):
+            return "oval"
+        elif isinstance(self.node, function.Split):
+            return "hexagon"
+        else:
+            return "box"
+
+    def __init__(self, node):
+        self.node = node
+        self.id_ = id(node)
+        self.attribute = {
+            "label": str(self.node),
+            "shape": self._shape()
+        }
+
+    def __str__(self):
+        attributes = ["%s=\"%s\"" % (k, v) for (k, v)
+                      in self.attribute.items()]
+        return "%s [%s];" % (self.id_, ",".join(attributes))
+
+
+class ComputationalGraph(object):
+    def __init__(self, edges):
+        self.edges = edges
+
+    def _to_dot(self):
+        ret = "digraph graphname{"
+        for edge in self.edges:
+            head, tail = edge
+            assert (isinstance(head, variable.Variable)
+                    and isinstance(tail, function.Function)) or \
+                   (isinstance(head, function.Function)
+                    and isinstance(tail, variable.Variable))
+            head_node = DotNode(head)
+            tail_node = DotNode(tail)
+            ret += head_node.label()
+            ret += tail_node.label()
+            ret += "%s -> %s;" % (head_node.id_, tail_node.id_)
+        ret += "}"
+        return ret
+
+    def dump(self, format='dot'):
+        if format == 'dot':
+            return self._to_dot()
+        else:
+            NotImplementedError
+
+    def __len__(self):
+        return len(self.edges)
+
+    def __contains__(self, e):
+        return e in self.edges
+
+
+def computational_graph(outputs, remove_split=False):
+    cands = []
+    seen_edges = set()
+
+    def heap_with_push_counter():
+        table = []
```

Comment:
It seems the table is used only for taking its length. How about just counting the number of adds?